lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -622,14 +625,43 @@ coinSelections =
 
 selectCoins
     :: forall ctx s t k n.
-        ( s ~ SeqState n k
+        ( Buildable (ErrValidateSelection t)
+        , DelegationAddress n k
+        , k ~ ShelleyKey
+        , s ~ SeqState n k
         , ctx ~ ApiLayer s t k
         )
     => ctx
     -> ApiT WalletId
     -> ApiSelectCoinsData n
     -> Handler (ApiCoinSelection n)
-selectCoins _ctx (ApiT _wid) _selectCoinsData = throwError err501
+selectCoins ctx (ApiT wid) body = do
+    CoinSelection mInputs mOutputs mChange <-
+        liftHandler $ withWorkerCtx ctx wid liftE $ \wrk ->
+            W.selectCoinsForPayment @_ @s @t wrk wid $
+                coerceCoin <$> (body ^. #payments)
+    mkApiCoinSelection
+        <$> (ensureNonEmptyInputs mInputs)
+        <*> (ensureNonEmptyOutputs =<<
+                liftHandler (withWorkerCtx ctx wid throwE $ \wrk ->
+                    W.assignChangeAddressesAndCheckpoint
+                        @_ @s @k wrk wid () mOutputs mChange))
```

Comment:
**[suggestion]**

Swap the use of applicative for a monadic approach. 

```hs
let payments = coerceCoin <$> (body ^. #payments)
CoinSelection mInputs mRecipients mChange <- liftHandler 
    $ withWorkerCtx ctx wid liftE 
    $ \wrk -> W.selectCoinsForPayment @_ @s @t wrk wid payments

mOutputs  <- liftHandler 
    $ withWorkerCtx ctx wid throwE 
    $ \wrk -> assignChangeAddressesAndCheckpoint @_ @s @k wrk wid () mRecipients mChange

neInputs  <- ensureNonEmpty ErrSelectCoinsUnableToAssignInputs mInputs
neOutputs <- ensureNonEmpty ErrSelectCoinsUnableToAssignOutputs mOutputs

pure $ mkApiCoinSelection neInputs neOutputs

```

**Why?**

(In this case) It leads to more readable code without changing much of the semantic.

Reply:
Hi @KtorZ. Thanks for this suggestion: I agree this could be more readable.

I've made the following change, in response to your suggestion: https://github.com/input-output-hk/cardano-wallet/pull/1165/commits/c812bf8126dc324e9ebf2fbf1ff1fe0fe960c4b1.